core/src/main/java/io/grpc/internal/DnsNameResolver.java

```diff
@@ -226,4 +237,163 @@ public final synchronized void shutdown() {
+    }
+  }
+
+  /**
+   * A composite DNS resolver that uses both the JDK and JNDI resolvers as delegate.  It is
+   * expected that two DNS queries will be executed, with the second one being from JNDI.  This
+   * /should/ be cached, hopefully avoiding the second RTT to get the TXT records.  From manual
+   * testing, it seems that an ANY query is sent for both resolvers, which typically includes the
+   * TXT records.
+   */
+  @VisibleForTesting
+  static final class CompositeResolver extends DelegateResolver {
+
+    @Nullable private final DelegateResolver jdkResovler;
+    @Nullable private final DelegateResolver jndiResovler;
+
+    CompositeResolver(
+        @Nullable DelegateResolver jdkResovler, @Nullable DelegateResolver jndiResovler) {
+      this.jdkResovler = jdkResovler;
+      this.jndiResovler = jndiResovler;
+      checkArgument(jdkResovler != null || jndiResovler != null, "missing resolver");
+    }
+
+    @Override
+    ResolutionResults resolve(String host) throws Exception {
+      List<InetAddress> addresses = null;
+      List<String> txtRecords = null;
+      // Try to use the Jdk Resolver for addresses, which honors RFC 3484 address ordering (by
+      // merit of calling libc).  However, if not present, still try to use the JNDI results.
+      // However, JNDI can return TXT records, so it takes precedence for those.
+      if (jdkResovler != null) {
+        ResolutionResults results = jdkResovler.resolve(host);
+        addresses = results.addresses;
+        txtRecords = results.txtRecords;
+      }
+      if (jndiResovler != null) {
+        ResolutionResults results = jndiResovler.resolve(host);
+        addresses = MoreObjects.firstNonNull(addresses, results.addresses);
+        txtRecords = results.txtRecords;
+      }
+      return new ResolutionResults(addresses, txtRecords);
+    }
+  }
+
+  /**
+   * The default name resolver provided with the JDK.  This is unable to lookup TXT records, but
+   * provides address ordering sorted according to RFC 3484.  This is true on OpenJDK, because it
+   * in turn calls into libc which sorts addresses in order of reachability.
+   */
+  @VisibleForTesting
+  static final class JdkResolver extends DelegateResolver {
+
+    @Override
+    ResolutionResults resolve(String host) throws Exception {
+      return new ResolutionResults(
+          Arrays.asList(InetAddress.getAllByName(host)),
+          Collections.<String>emptyList());
+    }
+  }
+
+  /**
+   * A resolver that uses the JNDI interface.  This class is capable of looking up both addresses
+   * and text records, but does not provide ordering guarantees.  When the {@link JdkResolver} is
+   * not available, this can be used as a fallback.  Some results are better than no results.
+   */
+  @VisibleForTesting
+  static final class JndiResolver extends DelegateResolver {
+
+    private static final String[] rrTypes = new String[]{"A", "AAAA", "TXT"};
```

Comment:
Can we just request TXT? You mentioning that ANY was used makes me nervous, and it would be rare to use the A and AAAA here. Let's just KISS and only query for what we need.